### ISSUE TITLE
Implement save games

### ISSUE THREAD

**Issue by SWY1985** [SWY1985/CivOne#248]

**Comment by AlexFolland** [comment]
Documenting map-related saving bugs:

In a game started in CivOne, saving caused the following problems:

1. Cities offset 1 to the left
![image](https://cloud.githubusercontent.com/assets/601517/23550554/70d7a0f4-ffc6-11e6-9d44-e9fe0547c193.png)

2. Villages re-generated
3. Roads and irrigation offset 2 to the right
![image](https://cloud.githubusercontent.com/assets/601517/23550580/91adc452-ffc6-11e6-8d93-8df106a1d152.png)

4. Fog of war offset 2 to the left
![image](https://cloud.githubusercontent.com/assets/601517/23550728/5f4f3f8a-ffc7-11e6-89aa-c0d0847f4cad.png)

Not all cities were offset 1 to the left
Not all fog of war was offset 2 to the left
![image](https://cloud.githubusercontent.com/assets/601517/23550638/01d7cb6a-ffc7-11e6-8ab2-7d8a9a5e94e2.png)

Units remained where they were relative to the land

I should load this save in Civilization and check for discrepancies, as you mentioned that loading is currently buggy.

**Comment by AlexFolland** [comment]
Okay, that save crashes the original Civilization.

![image](https://cloud.githubusercontent.com/assets/601517/23551031/037b366c-ffc9-11e6-81b4-5093d1c22477.png)

**Comment by SWY1985** [comment]
Yeah, the save game function is very, very unstable at the moment. First of all, I'm missing all kinds of data from the original game, making it nearly impossible to make anything out of it for the original game. Then I've got a lot of details in the save game wrong.
I mean to get it working and fully compatible with the original game soon though... thanks for reporting the issues, it will help me track down some of the bugs.

**Comment by kazgaroth011** [comment]
I also noticed one one occasion that after loading a saved game the huts (villages) respawn.

**Comment by SWY1985** [comment]
That's because I don't (yet) properly record unit visited tiles. It should get fixed soon.

**Comment by AlexFolland** [comment]
Alternatively, if you're going to record visited tiles instead of simply recording the current game state, saving an initial game state and input performed since that state would allow replay playback.  That could be used as the custom non-Civilization-compatible save format also.  To load a game on a certain date, the engine would simply play back the input recording without animating it.

**Comment by SWY1985** [comment]
The data recorded for the game playback is a lot more compact. Here's an overview: https://forums.civfanatics.com/threads/sve-file-format.493581/#post-12383993

Eventually, recording the entire game in the new save game format is a pretty cool idea. I'm going to steal it. ;-)

**Comment by kazgaroth011** [comment]
A couple of issues when loading saved games:
- immediately after loading a game I get a "choose new government" question
- tax & science rates stuck at 0 % (and cannot be changed as in the pic below)

![image](https://user-images.githubusercontent.com/21365802/34393522-6659d15a-eb53-11e7-9a8d-69f906b9056c.png)
